lib/mongoid/clients/options.rb

```diff
@@ -21,10 +21,30 @@ module Options
       def with(options_or_context, &block)
         original_context = PersistenceContext.get(self)
         original_cluster = persistence_context.cluster
-        set_persistence_context(options_or_context)
-        yield self
-      ensure
-        clear_persistence_context(original_cluster, original_context)
+        if block_given?
+          begin
+            set_persistence_context(options_or_context)
+            yield self
+          ensure
+            clear_persistence_context(original_cluster, original_context)
+          end
+        else
+          # If #with is called on a Criteria we want to return a new criteria,
+          # but if it's called on a document, we want to return the same document
+          # since duping a document changes its _id.
+          if is_a?(Criteria)
+            crit = dup
+            crit.instance_variable_set("@original_cluster", original_cluster)
+            crit.instance_variable_set("@original_context", original_context)
```

Comment:
what happens if I chain multiple non-block calls to `#with` (`Band.with(...).with(...).all`)? Does `@original_cluster` etc get set to the second-to-last call to `with`, or does it remember the state at the beginning of the Criteria?

Reply:
Interesting, the first `with` would probably be ignored tbh. I think this would be a separate feature however.